File: framework/languages/java/samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/gateway/GameSession.java
package systems.zlink.samples.zoneworld.server.gateway;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.framework.actors.ActorRef;
import systems.zlink.framework.actors.ZLinkActorCreateResult;
import systems.zlink.framework.actors.ZLinkActorManager;
import systems.zlink.framework.messaging.ZLinkMessage;
import systems.zlink.framework.streams.ZLinkSession;
import systems.zlink.framework.streams.ZLinkSessionContext;
import systems.zlink.framework.streams.ZLinkSessionDispatchContext;
import systems.zlink.framework.streams.ZLinkStreamError;
import systems.zlink.samples.zoneworld.shared.Messages;
import systems.zlink.samples.zoneworld.shared.ZoneWorldNames;
import systems.zlink.samples.zoneworld.shared.ZoneWorldSpec;
public final class GameSession implements ZLinkSession {
    private final ZLinkSessionContext context;
    private final ZLinkActorManager actors;

    public GameSession(
        ZLinkSessionContext context,
        ZLinkActorManager actors) {
        this.context = context;
        this.actors = actors;
    }

    @Override
    public ZLinkSessionContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onConnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDisconnected() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onError(ZLinkStreamError error) {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletionStage<Void> onDispatch(
        ZLinkSessionDispatchContext dispatch,
        ZLinkMessage payload) {
        if ("JoinWorldReq".equals(dispatch.packetName())) {
            return join(dispatch, payload);
        }
        if (context.actors().bound().size() != 1) {
            throw new IllegalStateException("JoinWorldReq must bind an actor before game traffic");
        }
        return context.actors().bound().get(0).relay(dispatch, payload);
    }

    private CompletionStage<Void> join(
        ZLinkSessionDispatchContext dispatch,
        ZLinkMessage payload) {
        Messages.JoinWorldReq request = payload.decode(Messages.JoinWorldReq.class);
        return actors.getOrCreate(request.playerId(), ZoneWorldNames.PLAYER_ACTOR_TYPE)
            .inMesh(ZoneWorldNames.MESH)
            .request(new Messages.EnterWorldReq(
                ZoneWorldSpec.SPAWN_X,
                ZoneWorldSpec.SPAWN_Y,
                false,
                0,
                0))
            .submit()
            .thenCompose(result -> {
                if (result instanceof ZLinkActorCreateResult.Rejected rejected) {
                    return CompletableFuture.failedFuture(new IllegalStateException(
                        rejected.reply().decode(Messages.EnterWorldRes.class).error()));
                }
                ActorRef actorRef = result instanceof ZLinkActorCreateResult.Created created
                    ? created.actor()
                    : ((ZLinkActorCreateResult.Existing) result).actor();
                return context.actors().bindOrGet(actorRef)
                    .thenCompose(actor -> actor.relay(dispatch, payload));
            });
    }
}

File: framework/languages/java/samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/server/gateway/GameSession.kt
package systems.zlink.samples.kotlin.zoneworld.server.gateway

import systems.zlink.framework.actors.ZLinkActorClient
import systems.zlink.framework.actors.ZLinkActorCreateResult
import systems.zlink.framework.actors.ZLinkActorManager
import systems.zlink.framework.kotlin.ZLinkSuspendingSession
import systems.zlink.framework.kotlin.await
import systems.zlink.framework.kotlin.decode
import systems.zlink.framework.kotlin.kotlin
import systems.zlink.framework.kotlin.requestToActor
import systems.zlink.framework.messaging.ZLinkMessage
import systems.zlink.framework.streams.ZLinkSessionContext
import systems.zlink.framework.streams.ZLinkSessionDispatchContext
import systems.zlink.framework.streams.ZLinkStreamError
import systems.zlink.samples.kotlin.zoneworld.shared.Messages
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldNames

class GameSession(
    private val sessionContext: ZLinkSessionContext,
    private val actors: ZLinkActorManager,
    private val actorClient: ZLinkActorClient,
    private val probes: RelocationProbeService,
) : ZLinkSuspendingSession() {
    private val kotlinActors = actors.kotlin()
    private val kotlinActorClient = actorClient.kotlin()

    override fun context() = sessionContext

    override suspend fun onConnectedSuspending() {}

    override suspend fun onDisconnectedSuspending() {}

    override suspend fun onErrorSuspending(error: ZLinkStreamError) {}

    override suspend fun onDispatchSuspending(
        dispatch: ZLinkSessionDispatchContext,
        payload: ZLinkMessage,
    ) {
        if (dispatch.packetName() == "RelocationPairReq") {
            sessionContext.client().kotlin().reply(probes.selectPair().await()).await()
            return
        }
        if (dispatch.packetName() == "ActorLocationProbeReq") {
            val request = payload.decode<Messages.ActorLocationProbeReq>()
            sessionContext
                .client()
                .kotlin()
                .reply(probes.findActor(request.actorId).await())
                .await()
            return
        }
        if (dispatch.packetName() == "FreshActorProbeReq") {
            val request = payload.decode<Messages.FreshActorProbeReq>()
            sessionContext.client().kotlin().reply(probes.createFresh(request.actorId)).await()
            return
        }
        if (dispatch.packetName() == "MessageFollowProbeReq") {
            val request = payload.decode<Messages.MessageFollowProbeReq>()
            val response =
                kotlinActorClient
                    .requestToActor<Messages.MessageFollowProbeRes>(request.actorId, request)
                    .await()
            sessionContext.client().kotlin().reply(response).await()
            return
        }
        if (dispatch.packetName() == "MessageFollowProbeMsg") {
            val message = payload.decode<Messages.MessageFollowProbeMsg>()
            kotlinActorClient.sendToActor(message.actorId, message).await()
            return
        }
        if (dispatch.packetName() == "JoinWorldMsg") {
            join(dispatch, payload)
            return
        }
        require(sessionContext.actors().bound().size == 1) {
            "JoinWorldMsg must bind an actor first"
        }
        sessionContext.actors().bound().single().relay(dispatch, payload).await()
    }

    private suspend fun join(dispatch: ZLinkSessionDispatchContext, payload: ZLinkMessage) {
        val request = payload.decode<Messages.JoinWorldMsg>()
        // --8<-- [start:doc-zw-session-bind]
        val result =
            kotlinActors
                .getOrCreate(request.playerId, ZoneWorldNames.PLAYER_ACTOR_TYPE)
                .inMesh(ZoneWorldNames.MESH)
                .request(ZLinkMessage.empty())
                .await()
        if (result is ZLinkActorCreateResult.Rejected) error("actor creation rejected")
        val actor =
            when (result) {
                is ZLinkActorCreateResult.Created -> result.actor
                is ZLinkActorCreateResult.Existing -> result.actor
                is ZLinkActorCreateResult.Rejected -> error("unreachable")
            }
        sessionContext.actors().bindOrGet(actor).await().relay(dispatch, payload).await()
        // --8<-- [end:doc-zw-session-bind]
    }
}
